[jianrong7] iP

jianrong7 opened the pull request:
# Duke Chatbot 😸 

> "Insanity is doing the iP over and over again, but expecting different results." - [Some genius ](https://encrypted-tbn2.gstatic.com/licensed-image?q=tbn:ANd9GcSUPobOTWGGkexkI3nf8u8LFP65DSHycLqi52PJLG0IDhOHSHHPSs1iJUW09LLdUcQiKnoF93mTYz7tDaQ)probably

Tired of remembering what to do? Too many things to remember?

**FRET NOT!**

## Duke is here to save the day!

It is
- Text-based
- Easy to use
- Super convenient

All you need to do is to merge this PR and then download the latest release. Not only is it **free**, it works like a _charm_.

### Features

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Email reminders (coming soon)

### Technical stuff

1. Uses OOP
2. Uses gradle
3. Has unit testing
4. Has javadocs
~~5. Has GUI~~

### How to run it?

1. run `./gradlew run` in the terminal.
2. voila!


I don't know what else to write to fulfil the fenced code block.
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jianrong7 committed:
level-0

jianrong7 committed:
feat: added ability to echo user unless user says bye

jianrong7 committed:
feat: added ability to add and list tasks

jianrong7 committed:
feat: added logic to mark and unmark tasks

jianrong7 committed:
refactor: use switch-case instead of if-else statements and broke down each case to its own method

jianrong7 committed:
chore: added javadocs to Duke and gitignore .class files

jianrong7 committed:
test: added io redirection test

jianrong7 committed:
feat: handle errors using proper exceptions

jianrong7 committed:
feat: added delete feature and changed tasks to use ArrayList instead of array

jianrong7 committed:
feat: commands now use enums

jianrong7 committed:
feat: added feature to save data to local file

jianrong7 committed:
Merge branch 'branch-Level-7'

jianrong7 committed:
feat: dates are now stored as LocalDateTime objects

jianrong7 committed:
Merge branch 'branch-Level-8'

jianrong7 committed:
refactor: introduce Ui, Parser and TaskList classes for more oop

jianrong7 committed:
Merge branch 'A-MoreOOP'

jianrong7 committed:
chore: separate classes into packages

jianrong7 committed:
Merge branch 'A-Packages'

jianrong7 committed:
fix: add checkstyle and modify mainclass

jianrong7 committed:
Merge branch 'add-gradle-support'

jianrong7 committed:
fix: run duke using gradle

jianrong7 committed:
test: add testing for Task class

jianrong7 committed:
test: add test for TaskList class

jianrong7 committed:
test: fix TaskList and tests

jianrong7 committed:
test: add test for Parser class

jianrong7 committed:
Merge branch 'A-JUnit'

jianrong7 committed:
chore: add jar file

jianrong7 committed:
chore: added javadocs to classes

jianrong7 committed:
feat: add find command

jianrong7 committed:
Merge branch 'branch-A-JavaDoc'

jianrong7 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/utils/Parser.java
#	src/main/java/duke/utils/TaskList.java

jianrong7 committed:
feat: Add checkstyle

jianrong7 committed:
fix: Fix styling issues according to checkstyle

jianrong7 committed:
Merge branch 'A-Checkstyle'

JeremyYong128 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,117 @@
+
+            try {
+                switch (keyword) {
+                case BYE: {
```
Good that you didn't add indentation for case clauses!

JeremyYong128 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,177 @@
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
```
I liked that you separated the duke imports and the java imports, allows readers to understand the dependencies more easily.

JeremyYong128 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class Task {
+    private final String description;
+    private boolean isDone;
```
I like that you kept all class variables private

JeremyYong128 commented on src/main/java/duke/utils/Storage.java:
```diff
@@ -0,0 +1,68 @@
+        try {
+            FileWriter filewriter = new FileWriter(this.path);
+            for (Task task : tasks) {
+                filewriter.write(task.toSaveString() + "\n");
```
I like that you still wrapped the loop body in curly brackets despite there only being one line.

JeremyYong128 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+
+    /**
+     * Creates a Deadline object.
+     * @return The deadline of the task.
```
Not all the Javadoc comments have an empty line between the description and parameter section, you could consider standardising the format.

JeremyYong128 reviewed (approved):
Overall, I found the code easy to read. No issues found except for the small inconsistencies in Javadoc comments.

jianrong7 committed:
feat: added gui

jianrong7 committed:
Merge branch 'branch-Level-10'

jianrong7 committed:
feat: add varargs

jianrong7 committed:
Merge branch 'branch-A-Varargs'

jianrong7 committed:
chore: added javadocs

wxwern commented on src/main/java/duke/common/Command.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Represents a command.
+ */
+public class Command {
```
I like your naming choice for the `handle` methods in this `Command` class, so they sound like actions.

Since `Command` appears to be a [utility class](https://www.baeldung.com/java-helper-vs-utility-classes), how about making its methods `static` instead and removing the need for a constructor (making it `private Command() { ... }`)?

wxwern commented on src/main/java/duke/common/Message.java:
```diff
@@ -0,0 +1,17 @@
+    public static final String FIND_TASKS = "Here are the matching tasks in your list:\n";
+    public static final String INVALID_COMMAND = "I'm sorry, but I don't know what that means :-(";
+    public static final String TASK_STATUS = "\nNow you have %d tasks in the list.";
+}
```
I like this organisation of having a class with constant references.

wxwern commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,177 @@
+            throw new DukeMissingArgumentException("The start and end time of an event cannot be empty.");
+        }
+        return new String[] {split[0], split2[0], split2[1]};
+    }
```
I like that you used guard clauses, which makes it so that there's reduced nesting of code blocks and easier to read.

wxwern commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,177 @@
+        boolean isDone = split[1].equals("1");
+        String description = split[2];
+        Task task = null;
+
```
It's not immediately clear what the format is from this alone and what the code is trying to parse. How about adding a high level description of what's going on?

For example, you may minimally link the docs to `Task#toSaveString`, so it is clear that `parseTask` intends to parse a format that's output by said method.

wxwern commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+     * Returns the description of the task that is to be saved by Storage.
+     * @return The description of the task that is to be saved by Storage.
+     */
+    public String toSaveString() {
```
I notice that your `toSaveString` implementations do not sanitise user inputs. Hence, user data may be corrupted if a user inputs the `|` character in their task description. You might need to escape or prevent inputs with said character to fix this.

wxwern reviewed (commented):
Overall, the code is really clean and easy to read. I've not found any serious issues, though I've also included some suggestions you could look into.

jianrong7 committed:
feat: add assertions

Currently, the code does not make use of the `assert` keyword.
Adding the `assert` keyword will improve the code quality since it can
help us document important assumptions that should hold at various
points in the code.

jianrong7 committed:
fix: extract code from duke.java

Make code cleaner and more logical, practicing SLAP.

jianrong7 committed:
chore: cleanup code for parser.java and storage.java

Parser.java
Abstracted `formatter` to a larger scope which can be accessed when parsing both Deadline and Event tasks.

Storage.java
Removed `file` which was being initialised but not accessed in the code.

jianrong7 committed:
fix: already use streams, remove sout

Already using streams. Removed System.out.println

jianrong7 committed:
Merge pull request #1 from jianrong7/branch-A-Assertions

feat: add assertions

jianrong7 committed:
Merge branch 'master' into branch-A-CodeQuality

jianrong7 committed:
Merge pull request #2 from jianrong7/branch-A-CodeQuality

improve code quality

jianrong7 committed:
Merge branch 'master' into branch-A-Streams

jianrong7 committed:
Merge pull request #3 from jianrong7/branch-A-Streams

fix: already use streams, remove sout

jianrong7 committed:
feat: add github actions

jianrong7 committed:
Merge branch 'branch-A-CI'

jianrong7 committed:
feat: add tags to tasks

jianrong7 committed:
fix: fix tests caused by new tags feature

jianrong7 committed:
fix: remove whitespace

jianrong7 committed:
Merge pull request #4 from jianrong7/branch-C-Tagging

feat: add tags to tasks

jianrong7 committed:
feat: add a better gui by making circles rounded, making text wrap better and having different chat background colours

jianrong7 committed:
Merge branch 'branch-A-BetterGui'

jianrong7 committed:
fix DukeException in GUI

jianrong7 committed:
Merge branch 'branch-A-MoreErrorHandling'

jianrong7 committed:
fix checkstyle errors

jianrong7 committed:
fix tests

jianrong7 committed:
add personality

jianrong7 committed:
Merge branch 'branch-A-Personality'

jianrong7 committed:
added user guide and ui

jianrong7 committed:
Merge branch 'branch-A-UserGuide'

jianrong7 committed:
fix docs header

jianrong7 committed:
added pictures to user guide

jianrong7 committed:
update user guide with images

jianrong7 committed:
Update README.md
